### ISSUE TITLE
Some required headers are not included at `is_vector`, `is_vector_like` and `check_not_nan`

### ISSUE THREAD

**Issue by Licht-T** [stan-dev/math#1106]
## Description
While solving https://github.com/stan-dev/math/issues/1036, I found some header references are missing and `make test-headers` fails, e.g. the below `is_vector` template is not resolved as expected since `stan/math/prim/arr/meta/is_vector.hpp` is not included.
https://github.com/stan-dev/math/blob/502487c511594ccb93eb979df5b8fe163becb417/stan/math/prim/scal/meta/is_vector_like.hpp#L22
These problem are simply solved by including missing header like https://github.com/stan-dev/math/pull/1105/commits/f96fffd29ee55c4132ac82bd912714f069566a39, but we might consider the code structure for common methods.

## Example
`make test-headers` fails at #1105.

## Expected Output
`make test-headers` runs successfully.

#### Current Version:
v2.18.1

**Comment by bob-carpenter** [comment]
Thanks for reporting.  Which OS  and C++ compiler is this running on?  Those header tests should always be run automatically as part of continuous integration, so nothing should get merged without passing them.

**Comment by Licht-T** [comment]
@bob-carpenter The current header tests work in spite of missing headers. For example, just using [`sort_asc<double>`](https://github.com/stan-dev/math/blob/502487c511594ccb93eb979df5b8fe163becb417/stan/math/prim/arr/fun/sort_asc.hpp#L20) from another header (like this https://github.com/stan-dev/math/pull/1105/commits/994a218c2d391638f09e33a528cb54e03e01243c?utf8=%E2%9C%93&diff=unified#diff-3a2588c161d24c6149c9af79e8b7906aR27) makes its header test failure.

You can easily check by this patch.
```bash
(dev-stan) [math] g++ -std=c++1y -pthread -Wno-sign-compare     -O3  -I . -I lib/eigen_3.3.3 -I lib/boost_1.66.0 -I lib/sundials_3.1.0/include    -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION     -c -O0 -include stan/math/prim/arr/fun/sort_asc.hpp test/dummy.cpp -o /dev/null

In file included from ./stan/math/prim/scal/err/check_not_nan.hpp:8:0,
                 from ./stan/math/prim/arr/fun/sort_asc.hpp:4,
                 from <command-line>:0:
./stan/math/prim/scal/fun/value_of_rec.hpp: In instantiation of ‘double stan::math::value_of_rec(T) [with T = std::vector<double>]’:
./stan/math/prim/scal/err/check_not_nan.hpp:18:28:   required from ‘static void stan::math::{anonymous}::not_nan<T_y, is_vec>::check(const char*, const char*, const T_y&) [with T_y = std::vector<double>; bool is_vec = false]’
./stan/math/prim/scal/err/check_not_nan.hpp:53:50:   required from ‘void stan::math::check_not_nan(const char*, const char*, const T_y&) [with T_y = std::vector<double>]’
./stan/math/prim/arr/fun/sort_asc.hpp:20:53:   required from here
./stan/math/prim/scal/fun/value_of_rec.hpp:24:10: error: invalid static_cast from type ‘const std::vector<double>’ to type ‘double’
   return static_cast<double>(x);
```
```diff
diff --git a/stan/math/prim/arr/fun/sort_asc.hpp b/stan/math/prim/arr/fun/sort_asc.hpp
index efea236470..e36d1ac7e3 100644
--- a/stan/math/prim/arr/fun/sort_asc.hpp
+++ b/stan/math/prim/arr/fun/sort_asc.hpp
@@ -16,8 +16,7 @@ namespace math {
  * @return Vector in ascending order.
  * @throw std::domain_error If any of the values are NaN.
  */
-template <typename T>
-inline std::vector<T> sort_asc(std::vector<T> xs) {
+inline std::vector<double> sort_asc(std::vector<double> xs) {
   check_not_nan("sort_asc", "container argument", xs);
   std::sort(xs.begin(), xs.end());
   return xs;
```

**Comment by bob-carpenter** [comment]
I'm afraid I don't follow that last message.  Are you saying if we replace the templated sort-ascending with a double-specific one, the tests fail?  Why would we do that?

Also, I'm probably not the one to be tagging on this.

**Comment by Licht-T** [comment]
@bob-carpenter This is just an example. Sorry for my super-confusing example!

What I want to say are
- The [`sort_asc(std::vector<T> xs)`](https://github.com/stan-dev/math/blob/master/stan/math/prim/arr/fun/sort_asc.hpp#L21)'s argument is `vector` type, and it uses [`check_not_nan`](https://github.com/stan-dev/math/blob/master/stan/math/prim/scal/err/check_not_nan.hpp#L53) inside itself.
- But [`is_vector_like`](https://github.com/stan-dev/math/blob/master/stan/math/prim/scal/meta/is_vector_like.hpp#L4), [`check_not_nan`](https://github.com/stan-dev/math/blob/master/stan/math/prim/scal/err/check_not_nan.hpp#L53) uses this, is not designed for accepting `vector` because `is_vector_like` header does not include ["stan/math/prim/arr/meta/is_vector.hpp"](https://github.com/stan-dev/math/blob/master/stan/math/prim/arr/meta/is_vector.hpp).
- The header test, which is calling `sort_asc(std::vector<double> xs)` like #1105, fails since it cannot resolve `is_vector` for `vector` type.

Also,
- The current header tests is "just compiling a single header file with empty dummy"
- This does not make any evaluation of templates, e.g. this header test cannot check whether `sort_asc<double>` is compilable with single header "stan/math/prim/arr/fun/sort_asc.hpp" or not.

**Comment by bob-carpenter** [comment]
Thanks for the clarification.   In general, our `scal` implementations don't bring in `arr` because dependencies only run the other way (from `arr` to `scal`).   If you need to test `is_vector_like` with arrays, you need to explicitly bring in the `arr` implementation. 

Thus I think `sort_asc` should bring in everything it needs in `is_vector_like` and `check_not_nan` implementations. 

The second set of  bullet points is tricky---there's no general solution to the problem.  We try to deal with the actual template instantiations in other tests.  These header tests may not be doing much for us any more as we've switched to assuming clients will include a top-level include file that takes care of all the subtle ordering issues.

**Comment by syclik** [comment]
This was fixed with #1107. I verified today that it's not an issue with the code base now.

(@rok-cesnovar, thanks for letting me know this wasn't an issue.)
